ethexe/db/src/database.rs

```diff
@@ -43,52 +43,65 @@ use gprimitives::H256;
 use parity_scale_codec::{Decode, Encode};
 use std::collections::{BTreeMap, BTreeSet, VecDeque};
 
-// TODO (gsobol): because router address is a part of almost all keys, consider to use different db for each router.
-// TODO (gsobol): make separate structures for each key prefix. Each structure should have own method for key generation.
+// todo(playxe): possibly a better way to iterate prefixes, don't just do `H256::from_u64_be_low(prefix)`.
 #[repr(u64)]
-enum KeyPrefix {
-    /// BlockSmallData key prefix uses two keys: router address and block hash.
-    BlockSmallData = 0,
-    /// BlockEvents key prefix uses two keys: router address and block hash.
-    BlockEvents = 1,
-
-    /// BlockProgramStates key prefix uses two keys: router address and block hash.
-    BlockProgramStates = 2,
-    /// BlockOutcome key prefix uses two keys: router address and block hash.
-    BlockOutcome = 3,
-    /// BlockSchedule key prefix uses two keys: router address and block hash.
-    BlockSchedule = 4,
-
-    /// ProgramToCodeId key prefix uses two keys: router address and program id.
-    ProgramToCodeId = 5,
-    /// InstrumentedCode key prefix uses three keys: router address, runtime id and code id.
-    InstrumentedCode = 6,
-    /// CodeUploadInfo key prefix uses two keys: router address and code id.
-    CodeUploadInfo = 7,
-    /// CodeValid key prefix uses two keys: router address and code id.
-    CodeValid = 8,
-
-    /// SignedTransaction key prefix uses one key: transaction hash.
-    SignedTransaction = 9,
-
-    /// LatestComputedBlock key prefix uses one key: router address.
+enum Key {
+    BlockSmallData(H256) = 0,
+    BlockEvents(H256) = 1,
+    BlockProgramStates(H256) = 2,
+    BlockOutcome(H256) = 3,
+    BlockSchedule(H256) = 4,
+
+    ProgramToCodeId(ProgramId) = 5,
+    InstrumentedCode(u32, CodeId) = 6,
+    CodeUploadInfo(CodeId) = 7,
+    CodeValid(CodeId) = 8,
+
+    SignedTransaction(H256) = 9,
+
     LatestComputedBlock = 10,
-    /// LatestSyncedBlockHeight key prefix uses one key: router address.
     LatestSyncedBlockHeight = 11,
 }
 
-impl KeyPrefix {
-    fn prefix(self) -> [u8; 32] {
-        H256::from_low_u64_be(self as u64).0
-    }
-
-    fn one(self, key: impl AsRef<[u8]>) -> Vec<u8> {
-        [self.prefix().as_ref(), key.as_ref()].concat()
-    }
-
-    fn two(self, key1: impl AsRef<[u8]>, key2: impl AsRef<[u8]>) -> Vec<u8> {
-        let key = [key1.as_ref(), key2.as_ref()].concat();
-        self.one(key)
+impl Key {
+    fn to_bytes(self) -> Vec<u8> {
```

Comment:
impl `fn prefix(&self) -> Vec<u8>` working with representatives u64 and use it here as well